src/zulip/mod.rs

```diff
@@ -0,0 +1,191 @@
+mod api;
+
+use crate::team_api::TeamApi;
+use api::{ZulipApi, ZulipUserGroup};
+use rust_team_data::v1::ZulipGroupMember;
+
+use std::collections::BTreeMap;
+
+use failure::Error;
+
+pub(crate) fn run(token: String, team_api: &TeamApi, dry_run: bool) -> Result<(), Error> {
+    let zulip_api = ZulipApi::new(token, dry_run);
+    let user_group_definitions = get_user_group_definitions(team_api, &zulip_api)?;
+    let mut controller = ZulipController::new(zulip_api, dry_run)?;
+
+    for (name, members) in user_group_definitions {
+        controller.create_or_update_user_group(&name, &members)?;
+    }
+
+    Ok(())
+}
+
+/// Fetches the definitions of the user groups from the Team API
+fn get_user_group_definitions(
+    team_api: &TeamApi,
+    zulip_api: &ZulipApi,
+) -> Result<BTreeMap<String, Vec<usize>>, Error> {
+    let email_map = zulip_api
+        .get_users()?
+        .into_iter()
+        .map(|u| (u.email, u.user_id))
+        .collect::<BTreeMap<_, _>>();
+    let user_group_definitions = team_api
+        .get_zulip_groups()?
+        .groups
+        .into_iter()
+        .map(|(name, group)| {
+            let members = &group.members;
+            let member_ids = members
+                .iter()
+                .filter_map(|member| match member {
+                    ZulipGroupMember::Email(e) => {
+                        let id = email_map.get(e);
+                        if id.is_none() {
+                            log::warn!("no Zulip id found for '{}'", e);
+                        }
+                        id
+                    }
+                    ZulipGroupMember::Id(id) => Some(id),
```

Comment:
What happens when an invalid Zulip ID is present in the team repo? If that makes the request fail we'll want to also log and ignore here.

Reply:
If the Zulip ID is invaild and the request fails, the request failure will be bubbled up with `?` and logged in main

Reply:
Hmm, does that mean the whole synchronization would abort? We already retrieved a list of all Zulip users in that function, we could also check whether the ID exists in the `filter_map` and warn plus return `None` if the ID is not valid.

Reply:
Yea, I think I'd also like to add a validation step for checking this info against Zulip in the team repo so that the ID is more than likely correct, but I suppose even if we validate, we can still run into race conditions where the id is not correct.

Reply:
I made a change. Not sure I really like how it's being handled, but it's probably a decent enough start...